mod/wikirate/set/type/source_import_file.rb

```diff
@@ -0,0 +1,228 @@
+include_set Type::File
+include_set Abstract::Import
+
+attachment :source_import_file, uploader: CarrierWave::FileCardUploader
+
+event :import_csv, :prepare_to_store,
+      on: :update,
+      when: proc { Env.params["is_source_import_update"] == "true" } do
+  return unless (sources = Env.params[:sources])
+  return unless valid_import_format?(sources)
+  source_map = {}
+  success.params[:slot] = {
+    updated_sources: [],
+    duplicated_sources: []
+  }
+  sources.each do |source|
+    source_card = parse_source_card source, source_map
+    if source_card
+      source_card.director.catch_up_to_stage :validate
+      source_card.director.transact_in_stage = :integrate
+    end
+    handle_import_errors source_card
+  end
+  clear_slot_params
+  handle_redirect
+end
+
+def source_args source_hash
+  args = super source_hash[:source]
+  args.merge(
+    "+title" => { content: Env.params[:title][source_hash[:row].to_s] },
+    "+report_type" => { content: "[[#{source_hash[:report_type]}]]" },
+    "+company" => { content: "[[#{source_hash[:company]}]]" },
+    "+year" => { content: "[[#{source_hash[:year]}]]" }
+  )
+end
+
+def create_source source_hash
+  source_card = add_subcard "", type_id: SourceID,
+                                subcards: source_args(source_hash)
+  Env.params[:sourcebox] = "true"
+  source_card.director.catch_up_to_stage :prepare_to_store
+  Env.params[:sourcebox] = nil
+  unless source_card.errors.empty?
+    source_card.errors.each { |k, v| errors.add k, v }
+  end
+  source_card
+end
+
+def create_or_update_subcard source_card, trait, content
+  trait_card = source_card.fetch trait: trait,
+                                 new: { content: "[[#{content}]]" }
```

Comment:
It is called for `report_type`, `wikirate_company` and `year` which all have the pointer type as default.

let me rename it to `create_or_update_pointer_subcard` to tell what it can do 😉